lib/xebow.ex

```diff
@@ -37,4 +41,159 @@ defmodule Xebow do
 
   @spec layout() :: Layout.t()
   def layout, do: @layout
+
+  @type animations :: [Animation.t()]
+  @type animation_params :: %{String.t() => atom | number | String.t()}
+
+  defmodule State do
+    @moduledoc false
+    defstruct [:current_animation, :next_list, :previous_list]
+  end
+
+  use GenServer
+
+  # Client Implementations:
+
+  @doc """
+  Starts the Xebow application, which manages initialization of animations, as well
+  as switching between active animations. It also maintains animation config state
+  and persists it in memory between animation changes.
+  """
+  @spec start_link([]) :: GenServer.on_start()
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, Animation.types(), name: __MODULE__)
+  end
+
+  @doc """
+  Gets the animation configuration. This retrievs current values, which allows for
+  changes to be made with `update_animation_config/1`
+  """
+  @spec get_animation_config() :: {Animation.Config.t(), keyword(Animation.Config.t())}
+  def get_animation_config do
+    GenServer.call(__MODULE__, :get_animation_config)
+  end
+
+  @doc """
+  Switches to the next active animation
+  """
+  @spec next_animation() :: :ok
+  def next_animation do
+    GenServer.cast(__MODULE__, :next_animation)
+  end
+
+  @doc """
+  Switches to the previous active animation
+  """
+  @spec previous_animation() :: :ok
+  def previous_animation do
+    GenServer.cast(__MODULE__, :previous_animation)
+  end
+
+  @doc """
+  Updates the configuration for the current animation
+  """
+  @spec update_animation_config(animation_params) ::
+          :ok
+  def update_animation_config(params) do
+    GenServer.cast(__MODULE__, {:update_animation_config, params})
+  end
+
+  # Server Implementations:
+
+  @impl GenServer
+  def init(types) do
+    active_animations =
+      types
+      |> Enum.map(&initialize_animation/1)
+
+    current = hd(active_animations)
+    Engine.set_animation(current)
+
+    state = %State{
+      current_animation: current,
+      next_list: active_animations,
+      previous_list: []
+    }
+
+    {:ok, state}
+  end
+
+  @impl GenServer
+  def handle_call(:get_animation_config, _from, state) do
+    {:reply, Animation.get_config(state.current_animation), state}
+  end
+
+  @impl GenServer
+  def handle_cast(:next_animation, state) do
+    case state.next_list do
+      [] ->
+        [new_current | next_list] = Enum.reverse([state.current_animation | state.previous_list])
+        Engine.set_animation(new_current)
+
+        state = %{
+          state
+          | current_animation: new_current,
+            next_list: next_list,
+            previous_list: []
```

Comment:
:information_source: (optional)

Rather than storing `next_list` and `previous_list`, you could store a single list if the current animation were at the head.

_(I prefer the cursor-based approach more, so this recommendation is a second choice if you proceed with the list route.)_

Cycling forwards could then be like this:

```ex
[previous | rest] = animations
animations = rest ++ [previous]
current = hd(animations)
```

And backwards like this:

```ex
{current, rest} = List.pop_at(animations, -1)
animations = [current | rest]
```

This avoids the issue of having to detect if you've reached the end of a list, since this will cycle indefinitely.

Reply:
What I'm trying to do is make this operation O(1). Both the list + cursor and the list + append methods are O(N).

I had an idea last night, though, so let me get it working first, then I'll see what you think of it.

Reply:
Sorry, I didn't follow how big-O applies to cycling through animations. Could you explain how this animation data is going to scale up and cause performance problems?